Update to account for new components (#14588)
pulumi/docs: content/docs/iac/concepts/resources/components.md

- /docs/concepts/resources/components/
---

> Pulumi Components enable you to create, share, and consume reusable infrastructure building blocks across your organization and the broader community. Components instantiate a set of related resources, actng as an abstraction to encapsulates the resources' implementation details.

Reviewer: ```suggestion
Pulumi Components enable you to create, share, and consume reusable infrastructure building blocks across your organization and the broader community. Components instantiate a set of related resources, acting as an abstraction to encapsulate the resources' implementation details.
```

{{< chooser language "typescript,python,go,csharp,java" >}}

{{% choosable language typescript %}}

```typescript
runtime: nodejs
```

{{% /choosable %}}
{{% choosable language python %}}

```python
runtime: python
```

{{% /choosable %}}
{{% choosable language go %}}

```go
runtime: go
```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
runtime: dotnet
```

{{% /choosable %}}
{{% choosable language java %}}

```java
runtime: java
```

{{% /choosable %}}

{{< /chooser >}}

### Define an Entry Point

The entrypoint analyzes components to automatically build a schema, and interact with the Pulumi engine to mange the component lifecycle.

{{< chooser language "typescript,python,go,csharp,java" >}}

{{% choosable language typescript %}}

Not required for TypeScript.

{{% /choosable %}}
{{% choosable language python %}}

1. Create a `_main_.py` file in your component directory
2. In the `main` function, add a call to `component_provider_host`, specifying a list of components for the `components` argument

```python
from pulumi.provider.experimental import Metadata, component_provider_host
from staticpage import StaticPage

if __name__ == "__main__":

>     component_provider_host(name="python-components", components=[MyComponent])

Reviewer 2: Does `MyComponent` also need to be imported here?

>     component_provider_host(name="python-components", components=[MyComponent])

Reviewer: ```suggestion
    component_provider_host(name="python-components", components=[StaticPage])
```

>     component_provider_host(metadata=Metadata("python-components"), components=[MyComponent])

Reviewer: ```suggestion
    component_provider_host(name="python-components", components=[MyComponent])
```

```

{{% /choosable %}}
{{% choosable language go %}}

1. Define a `main.go` file

> 2. Declare an instance of `NewProviderBuilder`,  passing in a name space and the components being built

Reviewer: ```suggestion
2. Declare an instance of `NewProviderBuilder`,  passing in a name, namespace and the components being built
```

```go
package main

import (
    "github.com/pulumi/pulumi-go-provider/infer"
)

func main() {
    err := infer.NewProviderBuilder().
            WithName("go-components").
            WithNamespace("your-org-name").
            WithComponents(
                infer.Component(MyComponent),
            ).
            BuildAndRun()

>     if err != nil {
>         panic(err)
>         }
>     }

Reviewer 2: ```suggestion
    if err != nil {
        panic(err)
    }
}
```

```

{{% /choosable %}}
{{% choosable language csharp %}}

1. Create a `Program.cs` file
2. Add an entry point that calls the `ComponentProviderHost`

```csharp
using System.Threading.Tasks;

class Program
{
    public static Task Main(string []args) =>
        Pulumi.Experimental.Provider.ComponentProviderHost.Serve(args);
}
```

{{% /choosable %}}
{{% choosable language java %}}

1. Create an `App.java` file
2. Create a new instance of `ComponentProviderHost` in the entry point

```java

> package io.mikhail.components;

Reviewer 2: ```suggestion
package com.example.components;
```

import java.io.IOException;
import com.pulumi.provider.internal.Metadata;
import com.pulumi.provider.internal.ComponentProviderHost;

public class App {
    public static void main(String[] args) throws IOException, InterruptedException {
        new ComponentProviderHost("java-components", App.class.getPackage()).start(args);
    }
}
```

{{% /choosable %}}

{{< /chooser >}}

### Publishing the Component

> Once a component is authored, it can be pushed to a git repo and consumed remotely; or, in situations like monorepos, the comonpent can be referenced locally.

Reviewer 2: ```suggestion
Once a component is authored, it can be pushed to a git repo and consumed remotely; or, in situations like monorepos, the component can be referenced locally.
```

> Once a component is authored, it can be pushed to a git repo and consumed remotely; or, in situations like monorepos, the comonpent can be referenced via locally.

Reviewer: ```suggestion
Once a component is authored, it can be pushed to a git repo and consumed remotely; or, in situations like monorepos, the comonpent can be referenced locally.
```

> Once a component is authored, it can be pushed to a git repo and consumed remotely; or, in situations like monorepos, the comonpent can be referenced via locally.

Reviewer 2: ```suggestion
Once a component is authored, it can be pushed to a git repo and consumed remotely; or, in situations like monorepos, the component can be referenced locally.
```

#### Git Consumption

> In the consuming Pulumi application, add the component as a depdency.

Reviewer: ```suggestion
In the consuming Pulumi application, add the component as a dependency.
```

```bash
pulumi package add github.com/myorg/my-component
```

> If you're using version tags, you can specify those as well

Reviewer: ```suggestion
If you're using version tags, you can specify those as well.
```
